tools/ruby-gems/udb/lib/udb/cli/create.rb

```diff
@@ -0,0 +1,689 @@
+        loop do
+          variable_location = prompt.ask \
+            "What is the location of the '#{variable_name}' variable (e.g., '5', '14-12', '31|7|30-25|11-8'. Type 'help' for more)",
+            required: true
+
+          if variable_location == "help"
+            location_help = <<~HELP
+              Locations can be specified as:
+                * A single digit, when the field is one bit wide
+                  ex: 5
+                  ex: 31
+                * A range of digits separated by a dash ('-'), with the MSB first
+                  ex: 14-12
+                  ex: 24-20
+                * A concatenated list of digits and/or ranges joined by a pipe ('|')
+                  Used when the field is split in the encoding.
+                  The bits of the field must be listed in order from MSB to LSB _of the field_ (not of the encoding)
+                    ex: 31|7|30-25|11-8  # location of 12-bit imm in B-Type instructions;
+                                         # imm[11] is at $encoding[31]
+                                         # imm[10] is at $encoding[7]
+                                         # imm[9:4] is at $encoding[30:25]
+                                         # imm[3:0] is at $encoding[11:8]
+
+                Press any key to continue
+            HELP
+
+            TtyTools.display_box(prompt, "Variable location", location_help)
+          else
+            break
+          end
+        end
+
+        variables << OpcodeOrVariable.new(name: variable_name, location: T.must(variable_location))
+      end
+
+      template_path = Udb.gem_path / "lib" / "udb" / "templates" / "instruction_type.yaml.erb"
+      dest_path = outdir / "inst_type" / "#{name}.yaml"
+
+      erb = ERB.new(template_path.read, trim_mode: "-")
+      erb.filename = template_path.to_s
+
+      CreatedFile.new(path: dest_path, contents: erb.result(binding), next_steps: [])
+    end
+
+    sig { params(prompt: TTY::Prompt, copyright: String, outdir: Pathname, inst_type: T.nilable(String)).returns(CreatedFile) }
+    def self.create_inst_var(prompt, copyright, outdir, inst_type)
+      name = prompt.ask \
+        "What is the instruction var name (e.g., 'xs1')?",
+        modify: [:trim]
+
+      inst_types =
+        Dir[Udb.repo_root / "spec" / "std" / "isa" / "inst_type" / "*.yaml"].map do |f|
+          next if !inst_type.nil? && inst_type != File.basename(f, ".yaml")
+          { name: File.basename(f, ".yaml"), contents: YAML.load_file(f) }
+        end.compact
+      raise if inst_types.empty?
+
+      opcode_fields = {}
+      inst_types.each do |itype|
+        itype[:contents]["variables"].each_key do |itype_var_name|
+          opcode_fields["#{itype[:name]}:#{itype_var_name}"] =
+            { type: itype, var: itype_var_name }
+        end
+      end
+
+      opcode_field = prompt.select \
+        "What opcode field does this variable correspond with?",
+        opcode_fields.keys
```

Comment:
so this is applying a variable name to a defined bit field in the opcode subtype?